### PR TITLE
[SRVKE-1380] Fix Consumer finalization 

### PR DESCRIPTION
**Pull request by pierDipi**
When we use the DeleteResource function we check for changed == coreconfig.ResourceChanged to then proceed to update the CM, however, by returning idx, changed == coreconfig.ResourceChanged is almost never true, so we never update the CM with the resource removed.

Tested in https://github.com/openshift-knative/eventing-kafka-broker/pull/547 without the fix

### CODE REVIEW

File: control-plane/pkg/reconciler/consumer/consumer.go [hunk 1]
```diff
@@ -307,16 +307,16 @@ func addResource(resource *contract.Resource) contractMutatorFunc {
 func removeResource(_ *zap.Logger, ct *contract.Contract, c *kafkainternals.Consumer) int {
 	idx := coreconfig.FindResource(ct, c.GetUID())
 	if idx == coreconfig.NoResource {
-		return coreconfig.NoResource
+		return coreconfig.ResourceUnchanged
 	}
 	coreconfig.DeleteResource(ct, idx)
-	return idx
+	return coreconfig.ResourceChanged
```
Comment (pierDipi) [inline control-plane/pkg/reconciler/consumer/consumer.go:313]: I believe this was the reason.

when we use this function we check for `changed == coreconfig.ResourceChanged` to then proceed to update the CM, however, by returning `idx`, `changed == coreconfig.ResourceChanged` is almost never true, so we never update the CM with the resource removed.

File: test/e2e_new/kafka_source_test.go [hunk 2]
```diff
@@ -44,3 +45,25 @@ func TestKafkaSourceCreateSecretsAfterKafkaSource(t *testing.T) {
...
+			knative.WithKnativeNamespace(system.Namespace()),
+			knative.WithLoggingConfig,
+			knative.WithTracingConfig,
+			k8s.WithEventListener,
+			environment.InNamespace(namespace),
+			environment.Managed(t),
+		)
+
+		env.Test(ctx, t, features.SetupNamespace(namespace))
+		env.Test(ctx, t, features.SetupAndCleanupKafkaSources(50))
```
Comment (matzew) [inline test/e2e_new/kafka_source_test.go:67]: 50 per loop seems to be a bit more than the actual reported case. :+1:

File: control-plane/pkg/reconciler/consumer/consumer.go (hunk 1 above)
Comment (matzew) [reply control-plane/pkg/reconciler/consumer/consumer.go:313]: seems reasonable, perhaps to add a comment?

Thanks for the reasoning here

Comment (matzew) [reply control-plane/pkg/reconciler/consumer/consumer.go:313]: FWIW, we can add a comment for the upstream PR... just saying (if we want to ...)

File: test/e2e_new/features/kafka_source.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+func SetupAndCleanupKafkaSources(n int) *feature.Feature {
+	f := feature.NewFeatureNamed("setup and cleanup KafkaSources")
+
+	prefix := "kafka-source-"
+	sink := "sink"
+
+	f.Setup("install a sink", svc.Install(sink, "app", "rekt"))
+
+	for i := 0; i < n; i++ {
+		topicName := feature.MakeRandomK8sName("topic") // A k8s name is also a valid topic name.
```
Comment (maschmid) [inline test/e2e_new/features/kafka_source.go:39]: is the kafkatopic being created anywhere?

Comment (pierDipi) [reply test/e2e_new/features/kafka_source.go:39]: No, good catch!

**Comment by maschmid** [comment]
test fails with

```     service.go:51: services "sink" is forbidden: unable to create new content in namespace kafka-source-deleted-recreated because it is being terminated ```

not sure where it's being deleted , perhaps by `environment.Managed`

**Comment by matzew** [comment]
```
  *[2023-01-11T15:23:02Z]NoClusters: No clusters in pool are ready to be claimed 
```

/test 411-test-reconciler-aws-ocp-411

**Pull request merged**